src/app/actions.ts

```diff
@@ -19,19 +20,28 @@ export type RegionTree = Array<{
   }>;
 }>;
 
+const topLevelFuseOptions = {
+  includeScore: true,
+  keys: ["name"],
+};
+const fuse = new Fuse(regionCodes, topLevelFuseOptions);
+
 function findAllRegionsByQuery(query: string): Region[] | undefined {
   const queryLower = query.toLowerCase();
   const results: Region[] = [];
-  for (const region of regionCodes) {
-    if (region.name.toLowerCase().includes(queryLower)) {
-      results.push({ name: region.name, code: region.code });
-    }
 
-    if (region.subnational1) {
-      for (const sub1 of region.subnational1) {
+  for (const region of fuse.search(query)) {
```

Comment:
If I understand correctly, right now when we search via fuse it's returning every object _and_ all of it's nested regions (which might not match the search themselves but are nested inside a region that does match). 

We then loop over all of these and push them into a single array which we return to the client.

This means that we get results that don't necessarily make sense to the user, i.e:

<img width="1350" alt="image" src="https://github.com/user-attachments/assets/6df0ba8f-2c54-4141-938d-25d122ae8dc8">

I _think_ we can approach this from a few different directions. In order of preference we can:

1. Flatten our data such that's it's a single array of objects, with each object referring to a single region. This means we can do a search on the `name` and only get back results that match. We would likely have to the data in such a way to get a better "name" output for nested regions (e.g each subnational might also have a `fullHierarchyName` that's like `${subnational2.name}, ${subnational1.name}, ${country.name}`). This option requires a bit more up-front work but means that the resulting implementation is pretty simple (and fast)
2. We search through our entire dataset, then for each result if it has subnationals we create a new Fuse instance and search through _those_ with the same query. This would be _slow_ and ergonomically weird in code.
3. We can trigger [multiple nested searches](https://www.fusejs.io/examples.html#nested-search) to look at specifically the top level countries, then the subnational1, then subnational2, then we can build a list of results from that after. This would be a bit faster than 3 but still slow and still ergonomically weird.

Reply:
I think option 1 is a very good shout, and your suggestion of the fullHierarchyName is along the lines of what I was thinking, in terms of the output to the user. i.e. you search "London" you get "London, England, United Kingdom" as per the subnational2, 1, country structure based on the original api data.

So in an example here:

```
name: "Argentina",
    code: "AR",
    subnational1: [
      {
        code: "AR-B",
        name: "Buenos Aires",
        subnational2: [
          {
            code: "AR-B-AA",
            name: "Adolfo Alsina",
          },
```
Working top to bottom, our flattened data would look more like:
```
{
  name: "Argentina",
  code: "AR",
},
  code: "AR-B",
  name: "Buenos Aires, Argentina",
},
{
  code: "AR-B-AA",
  name: "Adolfo Alsina, Buenos Aires, Argentina",
},
```
Would it be worth putting in a key-value pair in for subnational? 
name: ""
code: ""
subnational: "country" | "subnational1" | "subnational2"

Reply:
Yeah I like that. 

My only suggestion would be to store the original `name` and also some other `fullHierarchyName`. The reason being we can then decide in fuse which value to search on. 

With your proposed shape if a user were to search for "argentina" they would get back every city within argentina as well as the country itself. 

If we do something like: 

```
{
  name: "Argentina",
  code: "AR",
},
  code: "AR-B",
  name: "Buenos Aires",
  fullHierarchyName: "Buenos Aires, Argentina",
},
{
  code: "AR-B-AA",
  name: "Adolfo Alsina",
  fullHierarchyName: "Adolfo Alsina, Buenos Aires, Argentina",
},
```

We can choose to search on `name` or `fullHierarchyName`, or both but with different weights. So that we have some flexibility in the way search works (Also the client will get back both the exact name _and_ the fullHierarchyName so that if they don't need to show the full hierarchy they don't have to cut it up themselves.

Reply:
Ahhh yes, that's smart, makes total sense.

Reply:
Oh or alternatively we could do something like

```
[
  {
    name: "Argentina",
    country: "Argentina",
    subdivision1: null,
    subdivision2: null,
    type: 'country',
    code: "AR",
  },
  {
    code: "AR-B",
    name: "Buenos Aires",
    type: 'subdivision1',
    country: "Argentina",
    subdivision1: 'Buenos Aires',
    subdivision2: null,
    fullHierarchyName: "Buenos Aires, Argentina",
  },
  {
    code: "AR-B-AA",
    name: "Adolfo Alsina",
    type: 'subdivision2',
    country: "Argentina",
    subdivision1: 'Buenos Aires',
    subdivision2: 'Adolfo Alsina',
    fullHierarchyName: "Adolfo Alsina, Buenos Aires, Argentina",
  },
]
```

This tells the consumer what type each one is as well as all the pieces they need to understand the hierarchy. Which gives plenty of flexibility in how we might display that in a UI. 

e.g in this case you could list out `name` in bold but then show the other parts after it perhaps in a smaller font so it's less obvious but still there and helpful.